core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -315,6 +316,18 @@ private RexNode simplifyComparison(RexCall e, RexUnknownAs unknownAs) {
   // e must be a comparison (=, >, >=, <, <=, !=)
   private <C extends Comparable<C>> RexNode simplifyComparison(RexCall e,
       RexUnknownAs unknownAs, Class<C> clazz) {
+    List<RelDataType> types = RexUtil.types(e.operands);
+
+    RelDataType type = SqlTypeUtil.consistentType(rexBuilder.getTypeFactory(),
+        SqlOperandTypeChecker.Consistency.COMPARE, types);
+    if ((type != null)
+        && (type.getSqlTypeName() == SqlTypeName.TIMESTAMP)
```

Comment:
Does the issue only happen on TimeStamp? how about TimeStamp with local zone? or Date?

Reply:
Thanks for the comment. I don't think this issue is only for Timestamp. So, I generalized this and added it as a method in Comparison.